src/CPlusPlusMangle.cpp

```diff
@@ -133,7 +134,7 @@ struct QualsState {
     bool last_is_pointer{false};
 
     const Type &type;
-    const std::string base_mode;
+    const std::string &base_mode;
```

Comment:
Gotta say that having const& member vars makes me a little nervous; I know it's safe with proper usage, but will clang-tidy (etc) notice danger if these classes get refactored to have different lifespans?

Reply:
...and in fact, this specific change seems to have injected the name-mangling test failure; we are stored a reference to a *temporary* std::string in the instantiation on line 191. IMHO we should undo all changes of this sort unless there is also a clang-tidy check of some sort that will verify temporaries aren't ever used.

Reply:
This change was me being dumb, not clang-tidy - clang-tidy wanted to std::move into a std:string member. I'll see if there's a checker for it though, as that would be useful.

Reply:
There were a number of other cases where a member var was changed to a const-ref too; these should probably changed back (unless/until we find a robust checker)

Reply:
Those were all done manually by me after reading the source and checking if it the usage was scoped correctly. I just got tired and sloppy, so I'll recheck them. This idiom is already used throughout the compiler and provides a real performance benefit, so I don't want to back it out. E.g. I found a place where a pointless copy was being made of a large std::map.

Reply:
One possible safer idiom: have the things accept const pointers instead, so that it's obvious at the call-site that a non-owning reference is being taken, but then immediately dereference them in the constructor to store in the const ref member variable.

Reply:
```
struct MyHelper {
  const Foo &foo;
  MyHelper(const Foo *foo) : foo(*foo) {}
};

...

Foo foo;
MyHelper helper(&foo);
```

Reply:
The downside is that it's possible to pass nullptr, and the signature hints that it might be meaningful to do so.

Reply:
In the absence of a checker (I couldn't find one) I think this might be a good solution. You're not going to write &(some rvalue) at the call-site. LMK if you agree and I'll make the change.

Reply:
Actually belay that. This gets gross-looking in cases where you're just passing through an existing const-ref. Sigh. I'll keep thinking.

Reply:
Yeah, this is a place where C++ just doesn't have a good option (I presume Rust's borrow semantics would be the solution here). 
- Agreed that pointless copy of maps, large structs (etc) is bad, but for "small" things like std::string I'd err on the side of safety unless we have hard numbers showing the performance win
- Agreed that const pointers instead is a mixed blessing here

May I suggest that you decouple the other clang-tidy changes from this one (for now at least) since the others seem less controversial? There's no urgency to land this yet.